hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/jdbc/join/FullScanResultSetIterator.java

```diff
@@ -0,0 +1,123 @@
+/*
+ * Copyright 2023 Hazelcast Inc.
+ *
+ * Licensed under the Hazelcast Community License (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://hazelcast.com/hazelcast-community-license
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.jet.sql.impl.connector.jdbc.join;
+
+import com.hazelcast.internal.nio.IOUtil;
+import com.hazelcast.logging.ILogger;
+import com.hazelcast.logging.Logger;
+import com.hazelcast.sql.HazelcastSqlException;
+
+import java.sql.Connection;
+import java.sql.PreparedStatement;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.util.Iterator;
+import java.util.NoSuchElementException;
+import java.util.function.Function;
+import java.util.function.Supplier;
+
+public class FullScanResultSetIterator<T> implements Iterator<T> {
+
+    private static final ILogger LOGGER = Logger.getLogger(FullScanResultSetIterator.class);
+    private Connection connection;
+    private final String sql;
+    private final Function<ResultSet, T> rowMapper;
+    private final Supplier<T> emptyResultSetMapper;
+    private boolean hasNext;
+    private ResultSet resultSet;
+    private PreparedStatement preparedStatement;
+    private boolean callEmptyResultMapper = true;
+
+    private T nextItem;
+
+    public FullScanResultSetIterator(Connection connection,
+                                     String sql,
+                                     Function<ResultSet, T> rowMapper,
+                                     Supplier<T> emptyResultSetMapper
+    ) {
+        this.connection = connection;
+        this.sql = sql;
+        this.rowMapper = rowMapper;
+        this.emptyResultSetMapper = emptyResultSetMapper;
+    }
+    @Override
+    public boolean hasNext() {
+        try {
+            lazyInit();
+            hasNext = false;
+            if (getNextItemFromRowMapper()) {
+                callEmptyResultMapper = false;
+                hasNext = true;
+            } else {
+                if (getNextItemFromEmptyResultSetMapper()) {
+                    hasNext = true;
+                }
+            }
+            if (!hasNext) {
+                close();
+            }
+            return hasNext;
+        } catch (SQLException sqlException) {
+            close();
+            throw new HazelcastSqlException("Error occurred while iterating ResultSet", sqlException);
+        }
+    }
```

Comment:
The logic is over-complicated and may be refactored with my suggested changes.

Reply:
rev : 92b8d23155c211872167387fd532e313878e08d6